### PR TITLE
Queue Challenges

### PR DESCRIPTION
**Pull request by lramirezsuarez**
## Queues
Made the challenges for QueueArray, QueueLinkedList, QueueRingBuffer and QueueStack.

### CODE REVIEW

File: Data Structures and Algorithms/Data Structures and Algorithms/Queue/QueueArray.swift [hunk 1]
```diff
@@ -36,3 +36,15 @@ extension QueueArray: CustomStringConvertible {
...
+
+// MARK: - Challenge 4 method
+extension QueueArray {
+    public mutating func reversed() -> QueueArray {
+        var reversedQueue = QueueArray()
+        
+        while !self.isEmpty {
+            reversedQueue.enqueue(storage.removeLast())
+        }
+        return reversedQueue // return a copy of the reversed queue
```
Comment (jlainog) [inline Data Structures and Algorithms/Data Structures and Algorithms/Queue/QueueArray.swift:70]: are you sure this is a copy? the removeLast does remove the value from the storage so at the end your storage will be empty

File: Data Structures and Algorithms/Data Structures and Algorithms/Queue/QueueRingBuffer.swift [hunk 2]
```diff
@@ -38,3 +44,12 @@ extension QueueRingBuffer: CustomStringConvertible {
         return String(describing: buffer)
     }
 }
+
+extension QueueRingBuffer {
+    public mutating func reversed(size: Int) -> QueueRingBuffer {
+        let bufferReversed = self.buffer.reversed()
```
Comment (jlainog) [inline Data Structures and Algorithms/Data Structures and Algorithms/Queue/QueueRingBuffer.swift:46]: for reversing the ring buffer you need to take account for where the read index is, the reversed() call you are doing will reverse all the buffer from index 0.

File: Data Structures and Algorithms/Data Structures and Algorithms/Queue/QueueStack.swift [hunk 3]
```diff
@@ -43,3 +47,13 @@ extension QueueStack: CustomStringConvertible {
         return String(describing: queue)
     }
 }
+
+extension QueueStack {
+    public mutating func reversed() -> QueueStack {
+        var reversedQueueStack = QueueStack()
+        
+        rightStack.forEach( { reversedQueueStack.enqueue($0) } )
```
Comment (jlainog) [inline Data Structures and Algorithms/Data Structures and Algorithms/Queue/QueueStack.swift]: are you sure theres nothing in the left stack when this call is done ?

File: Data Structures and Algorithms/Data Structures and AlgorithmsTests/Queue/QueueChallengeTests.swift [hunk 4]
```diff
@@ -44,13 +111,13 @@ class QueueChallengeTests: XCTestCase {
         XCTAssertEqual(player3, totalPlayers[3])
         XCTAssertEqual(player4, totalPlayers[4])
         
-        XCTAssertFalse(queue.isEmpty)
+        XCTAssertTrue(queue.isEmpty)
```
Comment (jlainog) [inline Data Structures and Algorithms/Data Structures and AlgorithmsTests/Queue/QueueChallengeTests.swift]: the queue should not be empty after a round of the game ends. on the contrary it should hold all the player again in the right order

File: Data Structures and Algorithms/Data Structures and Algorithms/Queue/QueueArray.swift (hunk 1 above)
Comment (lramirezsuarez) [reply Data Structures and Algorithms/Data Structures and Algorithms/Queue/QueueArray.swift:70]: Solved, using a copy of the QueueArray, and removing the last from that copy.

**Pull request merged**